numpyro/hmc_util.py

```diff
@@ -781,3 +782,81 @@ def single_chain_init(key, only_params=False):
+    :param int num_draws: number of draws from the merged posterior.
+    :param jax.random.PRNGKey rng: source of the randomness, defaults to `jax.random.PRNGKey(0)`.
+    :return: a collection of `num_draws` samples with the same data structure as each subposterior.
+    """
+    rng = random.PRNGKey(0) if rng is None else rng
+    # stack subposteriors
+    joined_subposteriors = tree_multimap(lambda *args: np.stack(args), *subposteriors)
+    # shape = num_subposteriors x num_samples x sample_shape
+    joined_subposteriors = vmap(vmap(lambda sample: ravel_pytree(sample)[0]))(joined_subposteriors)
+
+    # randomly gets num_draws from subposteriors
+    n_subs = len(subposteriors)
+    n_samples = tree_flatten(subposteriors[0])[0][0].shape[0]
+    draw_idxs = random.randint(rng, shape=(n_subs, num_draws), minval=0, maxval=n_samples)
+    joined_subposteriors = vmap(lambda x, idx: x[idx])(joined_subposteriors, draw_idxs)
+
+    # compute weights for each subposterior (ref: Section 3.1 of [1])
+    weights = vmap(lambda x: np.linalg.inv(_cov(x)))(joined_subposteriors)
+    normalized_weights = np.matmul(np.linalg.inv(np.sum(weights, axis=0)), weights)
+
+    # get weighted samples
+    samples_flat = np.einsum('ijk,ilk->lj', normalized_weights, joined_subposteriors)
+
+    # unravel_fn acts on 1 sample of a subposterior
+    _, unravel_fn = ravel_pytree(tree_map(lambda x: x[0], subposteriors[0]))
+    return vmap(lambda x: unravel_fn(x))(samples_flat)
+
+
+def parametric(subposteriors, num_draws, rng=None):
+    """
+    Merges subposteriors following (embarrassingly parallel) parametric Monte Carlo algorithm.
+
+    **References:**
+
+    1. *Asymptotically Exact, Embarrassingly Parallel MCMC*,
+       Willie Neiswanger, Chong Wang, Eric Xing
+
+    :param list subposteriors: a list which each element is a collection of samples.
+    :param int num_draws: number of draws from the merged posterior.
+    :param jax.random.PRNGKey rng: source of the randomness, defaults to `jax.random.PRNGKey(0)`.
+    :return: a collection of `num_draws` samples with the same data structure as each subposterior.
+    """
+    rng = random.PRNGKey(0) if rng is None else rng
+    joined_subposteriors = tree_multimap(lambda *args: np.stack(args), *subposteriors)
+    joined_subposteriors = vmap(vmap(lambda sample: ravel_pytree(sample)[0]))(joined_subposteriors)
+
+    weights = vmap(lambda x: np.linalg.inv(_cov(x)))(joined_subposteriors)
+    cov = np.linalg.inv(np.sum(weights, axis=0))
+    normalized_weights = np.matmul(cov, weights)
+
+    # comparing to consensus implementation, we compute weighted mean here
+    submeans = np.mean(joined_subposteriors, axis=1)
+    mean = np.einsum('ijk,ik->j', normalized_weights, submeans)
+    samples_flat = dist.MultivariateNormal(mean, cov).sample(rng, (num_draws,))
+
+    _, unravel_fn = ravel_pytree(tree_map(lambda x: x[0], subposteriors[0]))
+    return vmap(lambda x: unravel_fn(x))(samples_flat)
```

Comment:
why return only samples and not instead return the mean/cov as well?

Reply:
I think we can add an arg `return_draws` to get `mean`/`cov`. The reason to return draws is draws will be used to be transformed back (in flow hmc) to the constrained domain. Or to be used in predictions. But `mean`/`cov` can be useful when users change their mind and want to get more draws without having to rerun this method. So I'll add an option to achieve it.

Reply:
I just add an optional arg `return_params` for your convenience. If you have other suggestion, please let me know.